File: stripe/dotnet/StripeIntegration.cs
// DreamCobots — Stripe .NET Integration
//
// Provides Payment Intent creation, Checkout sessions, and webhook handling.
//
// Setup:
//   dotnet add package Stripe.net
//   export STRIPE_SECRET_KEY=sk_...
//   dotnet run
using System;
using System.IO;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Stripe;
using Stripe.Checkout;

namespace DreamCobots.StripeIntegration
{
    /// <summary>
    /// DreamCobots Stripe helper — wraps common Stripe operations used by all bots.
    /// </summary>
    public class StripeService
    {
        private static readonly string WebhookSecret =
            Environment.GetEnvironmentVariable("STRIPE_WEBHOOK_SECRET") ?? string.Empty;

        static StripeService()
        {
            StripeConfiguration.ApiKey = Environment.GetEnvironmentVariable("STRIPE_SECRET_KEY") ?? string.Empty;
        }

        // -------------------------------------------------------------------
        // Payment Intent
        // -------------------------------------------------------------------

        /// <summary>Create a Stripe PaymentIntent for a one-time charge.</summary>
        public async Task<PaymentIntent> CreatePaymentIntentAsync(long amountInCents, string currency = "usd")
        {
            var options = new PaymentIntentCreateOptions
            {
                Amount = amountInCents,
                Currency = currency,
                AutomaticPaymentMethods = new PaymentIntentAutomaticPaymentMethodsOptions
                {
                    Enabled = true,
                },
            };
            var service = new PaymentIntentService();
            return await service.CreateAsync(options);
        }

        // -------------------------------------------------------------------
        // Subscription
        // -------------------------------------------------------------------

        /// <summary>Create a customer and subscribe them to a price.</summary>
        public async Task<Subscription> CreateSubscriptionAsync(string email, string priceId)
        {
            var customerOptions = new CustomerCreateOptions { Email = email };
            var customerService = new CustomerService();
            var customer = await customerService.CreateAsync(customerOptions);

            var subOptions = new SubscriptionCreateOptions
            {
                Customer = customer.Id,
                Items = new System.Collections.Generic.List<SubscriptionItemOptions>
                {
                    new SubscriptionItemOptions { Price = priceId },
                },
                PaymentBehavior = "default_incomplete",
            };
            subOptions.AddExpand("latest_invoice.payment_intent");

            var subService = new SubscriptionService();
            return await subService.CreateAsync(subOptions);
        }

        /// <summary>Cancel a subscription immediately.</summary>
        public async Task<Subscription> CancelSubscriptionAsync(string subscriptionId)
        {
            var service = new SubscriptionService();
            return await service.CancelAsync(subscriptionId);
        }

        // -------------------------------------------------------------------
        // Checkout Session
        // -------------------------------------------------------------------

        /// <summary>Create a Stripe Checkout Session.</summary>
        public async Task<Session> CreateCheckoutSessionAsync(
            string priceId, string successUrl, string cancelUrl, string mode = "subscription")
        {
            var options = new SessionCreateOptions
            {
                PaymentMethodTypes = new System.Collections.Generic.List<string> { "card" },
                LineItems = new System.Collections.Generic.List<SessionLineItemOptions>
                {
                    new SessionLineItemOptions { Price = priceId, Quantity = 1 },
                },
                Mode = mode,
                SuccessUrl = successUrl,
                CancelUrl = cancelUrl,
            };
            var service = new SessionService();
            return await service.CreateAsync(options);
        }

        // -------------------------------------------------------------------
        // Webhook Handler
        // -------------------------------------------------------------------

        /// <summary>Parse and handle an incoming Stripe webhook event.</summary>
        public static async Task HandleWebhookAsync(HttpContext context)
        {
            var json = await new StreamReader(context.Request.Body).ReadToEndAsync();
            var sigHeader = context.Request.Headers["Stripe-Signature"].ToString();

            Event stripeEvent;
            if (!string.IsNullOrEmpty(WebhookSecret))
            {
                try
                {
                    stripeEvent = EventUtility.ConstructEvent(json, sigHeader, WebhookSecret);
                }
                catch (StripeException ex)
                {
                    Console.WriteLine($"Webhook signature verification failed: {ex.Message}");
                    context.Response.StatusCode = 400;
                    return;
                }
            }
            else
            {
                stripeEvent = EventUtility.ParseEvent(json);
            }

            Console.WriteLine($"Webhook event received: {stripeEvent.Type}");

            switch (stripeEvent.Type)
            {
                case Events.PaymentIntentSucceeded:
                    var pi = stripeEvent.Data.Object as PaymentIntent;
                    Console.WriteLine($"Payment succeeded: {pi?.Id}");
                    break;
                case Events.CheckoutSessionCompleted:
                    var session = stripeEvent.Data.Object as Session;
                    Console.WriteLine($"Checkout completed: {session?.Id}");
                    break;
                case Events.CustomerSubscriptionCreated:
                    Console.WriteLine("Subscription created");
                    break;
                case Events.CustomerSubscriptionDeleted:
                    Console.WriteLine("Subscription canceled");
                    break;
                default:
                    Console.WriteLine($"Unhandled event type: {stripeEvent.Type}");
                    break;
            }

            context.Response.StatusCode = 200;
            await context.Response.WriteAsync("{\"received\": true}");
        }
    }

    // -----------------------------------------------------------------------
    // Minimal ASP.NET Core host
    // -----------------------------------------------------------------------

    public class Program
    {
        public static async Task Main(string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);
            builder.Services.AddSingleton<StripeService>();

            var app = builder.Build();

            app.MapPost("/webhook", StripeService.HandleWebhookAsync);

            var port = Environment.GetEnvironmentVariable("PORT") ?? "5000";
            Console.WriteLine($"DreamCobots Stripe webhook listener running on port {port}");
            Console.WriteLine($"Test locally: stripe listen --forward-to localhost:{port}/webhook");

            await app.RunAsync($"http://0.0.0.0:{port}");
        }
    }
}

File: stripe/java/StripeIntegration.java
// DreamCobots — Stripe Java Integration
//
// Provides Payment Intent creation, subscription management, and webhook handling.
//
// Setup:
//   Add stripe-java to pom.xml (see pom.xml in this directory)
//   export STRIPE_SECRET_KEY=sk_...
//   mvn spring-boot:run
package com.dreamcobots.stripe;

import com.stripe.Stripe;
import com.stripe.exception.SignatureVerificationException;
import com.stripe.exception.StripeException;
import com.stripe.model.*;
import com.stripe.model.checkout.Session;
import com.stripe.net.Webhook;
import com.stripe.param.*;
import com.stripe.param.checkout.SessionCreateParams;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * DreamCobots Stripe helper — wraps common Stripe operations used by all bots.
 */
public class StripeIntegration {

    static {
        Stripe.apiKey = System.getenv("STRIPE_SECRET_KEY");
    }

    private static final String WEBHOOK_SECRET =
            System.getenv().getOrDefault("STRIPE_WEBHOOK_SECRET", "");

    // -----------------------------------------------------------------------
    // Payment Intent
    // -----------------------------------------------------------------------

    /**
     * Create a Stripe PaymentIntent for a one-time charge.
     *
     * @param amountInCents Amount in the smallest currency unit (e.g. cents).
     * @param currency      ISO 4217 currency code (e.g. {@code "usd"}).
     * @return Stripe {@link PaymentIntent}.
     * @throws StripeException on API error.
     */
    public PaymentIntent createPaymentIntent(long amountInCents, String currency)
            throws StripeException {
        PaymentIntentCreateParams params = PaymentIntentCreateParams.builder()
                .setAmount(amountInCents)
                .setCurrency(currency)
                .setAutomaticPaymentMethods(
                        PaymentIntentCreateParams.AutomaticPaymentMethods.builder()
                                .setEnabled(true)
                                .build())
                .build();
        return PaymentIntent.create(params);
    }

    // -----------------------------------------------------------------------
    // Subscriptions
    // -----------------------------------------------------------------------

    /**
     * Create a customer and subscribe them to a price.
     *
     * @param email   Customer e-mail address.
     * @param priceId Stripe Price ID.
     * @return Stripe {@link Subscription}.
     * @throws StripeException on API error.
     */
    public Subscription createSubscription(String email, String priceId)
            throws StripeException {
        CustomerCreateParams customerParams = CustomerCreateParams.builder()
                .setEmail(email)
                .build();
        Customer customer = Customer.create(customerParams);

        SubscriptionCreateParams subParams = SubscriptionCreateParams.builder()
                .setCustomer(customer.getId())
                .addItem(SubscriptionCreateParams.Item.builder()
                        .setPrice(priceId)
                        .build())
                .setPaymentBehavior(SubscriptionCreateParams.PaymentBehavior.DEFAULT_INCOMPLETE)
                .addAllExpand(Arrays.asList("latest_invoice.payment_intent"))
                .build();
        return Subscription.create(subParams);
    }

    /**
     * Cancel an existing subscription immediately.
     *
     * @param subscriptionId Stripe Subscription ID.
     * @return Cancelled {@link Subscription}.
     * @throws StripeException on API error.
     */
    public Subscription cancelSubscription(String subscriptionId) throws StripeException {
        Subscription subscription = Subscription.retrieve(subscriptionId);
        return subscription.cancel();
    }

    // -----------------------------------------------------------------------
    // Checkout Session
    // -----------------------------------------------------------------------

    /**
     * Create a Stripe Checkout Session.
     *
     * @param priceId    Stripe Price ID.
     * @param successUrl Redirect URL on success.
     * @param cancelUrl  Redirect URL on cancel.
     * @param mode       {@code "payment"} or {@code "subscription"}.
     * @return Stripe Checkout {@link Session}.
     * @throws StripeException on API error.
     */
    public Session createCheckoutSession(
            String priceId, String successUrl, String cancelUrl, String mode)
            throws StripeException {
        SessionCreateParams.Mode sessionMode = "subscription"
                .equalsIgnoreCase(mode)
                ? SessionCreateParams.Mode.SUBSCRIPTION
                : SessionCreateParams.Mode.PAYMENT;

        SessionCreateParams params = SessionCreateParams.builder()
                .addPaymentMethodType(SessionCreateParams.PaymentMethodType.CARD)
                .addLineItem(SessionCreateParams.LineItem.builder()
                        .setPrice(priceId)
                        .setQuantity(1L)
                        .build())
                .setMode(sessionMode)
                .setSuccessUrl(successUrl)
                .setCancelUrl(cancelUrl)
                .build();
        return Session.create(params);
    }

    // -----------------------------------------------------------------------
    // Webhook Handler (to be wired into your HTTP framework)
    // -----------------------------------------------------------------------

    /**
     * Parse and handle an incoming Stripe webhook event.
     *
     * @param payload   Raw request body bytes.
     * @param sigHeader Value of the {@code Stripe-Signature} header.
     */
    public void handleWebhookEvent(String payload, String sigHeader) {
        Event event;
        if (!WEBHOOK_SECRET.isEmpty()) {
            try {
                event = Webhook.constructEvent(payload, sigHeader, WEBHOOK_SECRET);
            } catch (SignatureVerificationException e) {
                System.err.println("Webhook signature verification failed: " + e.getMessage());
                return;
            }
        } else {
            // Fallback for local testing without a signing secret
            event = Event.GSON.fromJson(payload, Event.class);
        }

        System.out.println("Webhook event received: " + event.getType());
        switch (event.getType()) {
            case "payment_intent.succeeded":
                System.out.println("Payment succeeded: " +
                        event.getDataObjectDeserializer().getObject()
                                .map(o -> ((PaymentIntent) o).getId()).orElse("unknown"));
                break;
            case "checkout.session.completed":
                System.out.println("Checkout completed");
                break;
            case "customer.subscription.created":
                System.out.println("Subscription created");
                break;
            case "customer.subscription.deleted":
                System.out.println("Subscription canceled");
                break;
            default:
                System.out.println("Unhandled event type: " + event.getType());
        }
    }

    // -----------------------------------------------------------------------
    // Main (quick smoke-test)
    // -----------------------------------------------------------------------

    public static void main(String[] args) throws StripeException {
        StripeIntegration integration = new StripeIntegration();
        PaymentIntent pi = integration.createPaymentIntent(2500L, "usd");
        System.out.println("PaymentIntent created: " + pi.getId());
    }
}
